website/testCSSOrder.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const path = require('path');
+const fs = require('fs');
+
+/*
+This verifies CSS ordering on the Docusaurus site itself,
+
+There are multiple ways to provide some CSS to Docusaurus
+and Docusaurus should guarantee a consistent CSS ordering over time
+
+See also
+- https://github.com/facebook/docusaurus/issues/3678
+- https://github.com/facebook/docusaurus/pull/5987
+
+TODO we should probably add a real e2e test in core instead of using our own website?
+Current solution looks good-enough for now
+
+ */
+
+// TODO temporary, the current order is bad and we should change that
+const EXPECTED_MARKER_CLASSES = [
+  // Some class from Infima, in the order Infima declares them
+  // this ensures we don't mess-up and re-order css from other libs
+  '.markdown>h2',
+  '.button--outline.button--active',
+  ':root', // TODO should be first
+  '.DocSearch-Hit-content-wrapper', // TODO should not be there?
+  '.navbar__title',
+  '.test-marker-site-custom-css-shared-rule', // TODO should not be there!
```

Comment:
...why is that?

Reply:
Imagine the following Infima CSS:

```css
.ifm-full-width {
  width: "100%";
}
.ifm-half-width {
  width: "50%";
}

.ifm-item {
  border: solid;
}
```

Imagine this element:

```html
<div className="ifm-half-width ifm-item"/>
```

Now a site owner wants to override some CSS with custom site CSS:

```css

.ifm-item {
  width: "100%"
}
```

User should expect the item to be 100% width, with the following CSS:

```css
.ifm-full-width {
  width: "100%";
}

.ifm-half-width {
  width: "50%";
}

.ifm-item {
  border: solid;
}

.ifm-item {
  width: "100%";
}
```

Now it's not what happens in practice, and the override does not work, because CSS is optimized as:


```css
.ifm-full-width, 
.ifm-item {
  width: "100%";
}

.ifm-half-width {
  width: "50%";
}

.ifm-item {
  border: solid;
}
```

`.ifm-half-width` still wins as it's last in the stylesheet, and the CSS optimization messed-up with the intent of the user.

Does it make sense?

Reply:
Yes, I was just missing the click word "optimization" there. You mentioned optimization below and it all makes sense. I don't think there's any overriding here, so it shouldn't be problematic how we order them

Reply:
In our case, there might not be any overriding leading to such weird behaviors, but in userland, it might still happen. 

Don't we want to prevent that to happen? Remember custom site CSS is optimized too, so it might impact some users in the end 🤷‍♂️ And upgrading docusauurs means our own internal CSS also changes, which can lead to unexpected CSS ordering changes from one release to another (ie some beta.1 customCSS could be at the bottom, and with beta.2 this exact same custom CSS ends up being at the very top)

Reply:
I suggest you re-try with a class name with not completely overlapping rules (e.g. add an extra `content: "hey"`. As soon as the user is really attempting to override anything, the selector would go to the bottom. (But in fact in order for her to "override" anything she has to use the same selector as an existing one, instead of a new selector. Relative positions of distinct selectors don't matter; only their priorities do)

Reply:
> I suggest you re-try with a class name with not completely overlapping rules


this is already the case with `test-marker-site-custom-css-unique-rule`, probably not worth adding another entry 🤷‍♂️ 

> As soon as the user is really attempting to override anything, the selector would go to the bottom.

I don't know, using `width: 100%` is a legit use-case for me, and the CSS selector of a user does not need to contain multiple rules, so it's eligible for de-duplication, even if it may not be the most common use-case

> But in fact in order for her to "override" anything she has to use the same selector as an existing one, instead of a new selector. Relative positions of distinct selectors don't matter; only their specificities do

🤷‍♂️ not really, there are many ways to create selectors targeting the same element, where those elements can have the exact same specificity, and insertion order still matters in those cases.

`<div className="x y z"/>`

You can target it with `.x`, `.y` or `.z`, all have the same specificity, and class insertion order will matter.

Reply:
Consider your example with `.ifm-full-width` and `.ifm-item` containing the same rules. Does it matter if the user uses `className="ifm-full-width"` or `className="ifm-item"`?—No, they both lead to the same style being applied. Yes, Infima classes may get hoisted by a user class name, but if they are always perfectly identical in bahavior, why does it matter which one we use?

I think we should trust the optimization algorithm, fix the site CSS insertion first, and after that try to come up with an actual counter-case to break the CSS

Reply:
agree, these are edge cases and less important than the general insertion order